### PR TITLE
chore(HACBS-1370): populate release-service

### PR DESCRIPTION
**Pull request by scoheb**

### CODE REVIEW

File: book/release-service.md [hunk 1]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
 
 ![](../diagrams/release-service/hacbs-release-service-data-flow.jpg)
 
+## Terminology
+
+**Snapshot** -  The custom resource that contains the list of all Components of an Application with their Component Image digests. Once created, the list of Components with their images is immutable. The Integration service updates the status of the resource to reflect the testing outcome.
```
Comment (johnbieren) [inline book/release-service.md:33]: You have two spaces instead of 1 after the hyphen

File: book/release-service.md [hunk 2]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+**Release Pipeline** -  An instance of a Pipeline that is orchestrated by the Release Service.
```
Comment (johnbieren) [inline book/release-service.md:34]: Two spaces after the hyphen

File: book/release-service.md [hunk 3]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+**Bundle** - A tekton bundle (image) that contains Tekton resources
+**Development Workspace** - workspace shared by 1 or more team members when code is built and tested.
```
Comment (johnbieren) [inline book/release-service.md:36]: The first 3 start with a capital after the hyphen

File: book/release-service.md [hunk 4]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+**Managed Workspace** - workspace that is managed by an SRE or Release Engineering team that is controlled. This is where Release pipeline execute.
```
Comment (johnbieren) [inline book/release-service.md:37]: The first 3 start with a capital after the hyphen. I think "A controlled workspace that is managed by an SRE or Release Engineering team" reads better. Kind of sounds like the Release Engineering team is controlled as written. Also, `s/execute/executes/`

File: book/release-service.md [hunk 5]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+
+## Resources
+Below are the list of CRs that the Release service is responsible for interacting and performing operations:
```
Comment (johnbieren) [inline book/release-service.md:40]: `s/Below are/Below is/`

Comment (johnbieren) [reply book/release-service.md:40]: interacting with? instead of just interacting

File: book/release-service.md [hunk 6]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+
+### CREATE
+
+| Custom Resource            | When?                                                    | Why?                                                                                       |
+|----------------------------|----------------------------------------------------------|--------------------------------------------------------------------------------------------|
+| SnapshotEnvironmentBinding | When deploying the Snapshot to an Environment            | Create a SnapshotEnvironmentBinding for the Referenced Environment when one does not exist |
+| PipelineRuns               | Once a Release has been composed and ready for execution | To perform the steps in the Release Strategy                                               |
```
Comment (johnbieren) [inline book/release-service.md:47]: L46 is singular, this is plural. Should probably be consistent. I also think `and is ready for execution` is better

File: book/release-service.md [hunk 7]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+**Managed Workspace** - workspace that is managed by an SRE or Release Engineering team that is controlled. This is where Release pipeline execute.
+
+## Resources
+Below are the list of CRs that the Release service is responsible for interacting and performing operations:
+
+### CREATE
+
+| Custom Resource            | When?                                                    | Why?                                                                                       |
+|----------------------------|----------------------------------------------------------|--------------------------------------------------------------------------------------------|
+| SnapshotEnvironmentBinding | When deploying the Snapshot to an Environment            | Create a SnapshotEnvironmentBinding for the Referenced Environment when one does not exist |
```
Comment (johnbieren) [inline book/release-service.md:46]: Maybe add `To` before `Create` like in line 47?

File: book/release-service.md (hunk 4 above)
Comment (johnbieren) [reply book/release-service.md:37]: the p in Release Pipeline is capitalized seemingly everywhere else. Should probably be consistent

File: book/release-service.md [hunk 8]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+| ReleasePlan             | When starting to process a Release                     | To determine the target workspace to run the Release Pipeline            |
+| ReleasePlanAdmission    | When starting to process a Release                     | To determine the Release Strategy to use to compose the Release Pipeline |
+| Snapshot                | Creating and triggering a Release Pipeline             | To determine which components to release                                 |
+| ReleaseStrategy         | Before composing the Release Pipeline                  | To determine the bundle to use to run the Release Pipeline               |
+
+### UPDATE
+
+| Custom Resource            | When?                                                    | Why?                                                         |
+|----------------------------|----------------------------------------------------------|--------------------------------------------------------------|
+| SnapshotEnvironmentBinding | When deploying the Snapshot to an Environment            | Update exiting SnapshotEnvironmentBinding for a new Snapshot |
```
Comment (johnbieren) [inline book/release-service.md:63]: `s/exiting/existing/`

File: book/release-service.md [hunk 9]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+| Application & Component | When validating ReleasePlans and ReleasePlanAdmissions | To ensure data consistency                                               |
+| ReleasePlan             | When starting to process a Release                     | To determine the target workspace to run the Release Pipeline            |
+| ReleasePlanAdmission    | When starting to process a Release                     | To determine the Release Strategy to use to compose the Release Pipeline |
+| Snapshot                | Creating and triggering a Release Pipeline             | To determine which components to release                                 |
+| ReleaseStrategy         | Before composing the Release Pipeline                  | To determine the bundle to use to run the Release Pipeline               |
+
+### UPDATE
+
+| Custom Resource            | When?                                                    | Why?                                                         |
+|----------------------------|----------------------------------------------------------|--------------------------------------------------------------|
```
Comment (johnbieren) [inline book/release-service.md:62]: I think the tables read better when they have `To` in their `Why`s

File: book/release-service.md [hunk 10]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+| SnapshotEnvironmentBinding | When deploying the Snapshot to an Environment            | Update exiting SnapshotEnvironmentBinding for a new Snapshot |
+| Release                    | During the lifecycle of an attempt to release a Snapshot | Provide status for the execution of the Release Pipeline     |
+
+### WATCH
+
+| Custom Resource              | When?                                                   | Why?                                                            |
+|------------------------------|---------------------------------------------------------|-----------------------------------------------------------------|
+| PipelineRun                  | Once Pipeline created                                   | Relay the Release PipelineRun status to the Release for viewing |
+| SnapshotEnvironmentBinding   | After the SnapshotEnvironmentBinding created or updated | Relay the GitOps deployment status to the Release for viewing   |
```
Comment (johnbieren) [inline book/release-service.md:71]: After the SEB *is* created or updated

File: book/release-service.md [hunk 11]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+| ReleaseStrategy         | Before composing the Release Pipeline                  | To determine the bundle to use to run the Release Pipeline               |
+
+### UPDATE
+
+| Custom Resource            | When?                                                    | Why?                                                         |
+|----------------------------|----------------------------------------------------------|--------------------------------------------------------------|
+| SnapshotEnvironmentBinding | When deploying the Snapshot to an Environment            | Update exiting SnapshotEnvironmentBinding for a new Snapshot |
+| Release                    | During the lifecycle of an attempt to release a Snapshot | Provide status for the execution of the Release Pipeline     |
+
+### WATCH
```
Comment (johnbieren) [inline book/release-service.md:66]: We also watch Releases... not sure if this is redundant and should be left omitted though

File: book/release-service.md [hunk 12]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+|------------------------------|---------------------------------------------------------|-----------------------------------------------------------------|
+| PipelineRun                  | Once Pipeline created                                   | Relay the Release PipelineRun status to the Release for viewing |
+| SnapshotEnvironmentBinding   | After the SnapshotEnvironmentBinding created or updated | Relay the GitOps deployment status to the Release for viewing   |
+
+#### Annotations/Labels
+
+Following the [annotation guidelines](https://docs.google.com/document/d/1gyXM3pkKFFfxHZnopBi_53vREFWhwA0pFUuIhopDuEo/edit#), the Release Service sets the below annotations on PipelineRuns.
+```
+“pipelines.appstudio.openshift.io/type":     "release",
+"appstudio.openshift.io/snapshot":           “<snapshot name>”,
```
Comment (johnbieren) [inline book/release-service.md:78]: not sure about this one

File: book/release-service.md [hunk 13]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+“appstudio.openshift.io/component":          “<component name>”,
```
Comment (johnbieren) [inline book/release-service.md:79]: not sure about this one

File: book/release-service.md [hunk 14]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+“appstudio.openshift.io/application":        “<application name>”,
+"release.appstudio.openshift.io/name":       "<release name>",
+"release.appstudio.openshift.io/namespace":  "<managed workspace name>"
```
Comment (johnbieren) [inline book/release-service.md:82]: we use the release namespace for this value. the release is always in the dev workspace, not managed

File: book/release-service.md [hunk 15]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+```
+The Release service will copy the annotations and labels from the Release CR and append those to the Release PipelineRuns for traceability across the system per [Labels and Annotations for StoneSoup pipelines](https://docs.google.com/document/d/1fJq4LDakLfcAPvOOoxxZNWJ_cuQ1ew9jfBjWa-fEGLE/edit#) and [StoneSoup builds and tests PRs](https://docs.google.com/document/d/113XTplEWRM63aIzk7WwgLruUBu2O7xVy-Zd_U6yjYr0/edit#)
```
Comment (johnbieren) [inline book/release-service.md:84]: Missing a `.` at the end of this sentence

File: book/release-service.md [hunk 16]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+
+## Detailed Workflow
+
+1. A Release CR is created automatically by the Integration Service or manually by a user.
+   * The `spec.snaphot` should reference the Snapshot to release.
+   * The `spec.releasePlan` should reference the ReleasePlan to use.
+2. Find the matching ReleasePlanAdmission in the `spec.target` namespace specified in the ReleasePlan.
+3. Extract the `spec.releaseStrategy` from the ReleasePlanAdmission
+4. Extract the `spec.serviceAccount`, `spec.bundle`, `spec.pipeline` and `spec.policy` to use from the Release Strategy.
```
Comment (johnbieren) [inline book/release-service.md:93]: Based on consistency, I think it should be `ReleaseStrategy`, not `Release Strategy`

File: book/release-service.md [hunk 17]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+5. Create a Release PipelineRun using the above info.
+6. Watch Release PipelineRun and update Release `status` with progress and outcome.
+7. If ReleasePlanAdmission specified an `spec.environment`, then do the following:
```
Comment (johnbieren) [inline book/release-service.md:96]: probably `a` spec.environment, not `an`

File: book/release-service.md [hunk 18]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+   * Copy the Snapshot to the managed workspace
+   * Verify that the Application and Components have been copied to managed workspace.
+   * Create or update SnapshotEnvironmentBinding in the managed workspace.
+   * Watch SnapshotEnvironmentBinding to relay deployment status to Release CR `status`
```
Comment (johnbieren) [inline book/release-service.md:100]: missing `.` at end

File: book/release-service.md [hunk 19]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+| SnapshotEnvironmentBinding | When deploying the Snapshot to an Environment            | Create a SnapshotEnvironmentBinding for the Referenced Environment when one does not exist |
+| PipelineRuns               | Once a Release has been composed and ready for execution | To perform the steps in the Release Strategy                                               |
+
+### READ
+
+| Custom Resources        | When?                                                  | Why?                                                                     |
+|-------------------------|--------------------------------------------------------|--------------------------------------------------------------------------|
+| Application & Component | When validating ReleasePlans and ReleasePlanAdmissions | To ensure data consistency                                               |
+| ReleasePlan             | When starting to process a Release                     | To determine the target workspace to run the Release Pipeline            |
+| ReleasePlanAdmission    | When starting to process a Release                     | To determine the Release Strategy to use to compose the Release Pipeline |
```
Comment (johnbieren) [inline book/release-service.md:55]: for consistency, probably no space in Release Strategy. It is a CRD name

File: book/release-service.md (hunk 7 above)
Comment (johnbieren) [inline book/release-service.md:46]: not sure Referenced should be capitalized; seems we are doing just CRD names capitalized

File: book/release-service.md (hunk 1 above)
Comment (scoheb) [reply book/release-service.md]: done

File: book/release-service.md (hunk 2 above)
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md (hunk 3 above)
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md (hunk 4 above)
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md (hunk 5 above)
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md [hunk 20]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+1. A Release CR is created automatically by the Integration Service or manually by a user.
+   * The `spec.snaphot` should reference the Snapshot to release.
+   * The `spec.releasePlan` should reference the ReleasePlan to use.
+2. Find the matching ReleasePlanAdmission in the `spec.target` namespace specified in the ReleasePlan.
+3. Extract the `spec.releaseStrategy` from the ReleasePlanAdmission
+4. Extract the `spec.serviceAccount`, `spec.bundle`, `spec.pipeline` and `spec.policy` to use from the Release Strategy.
+5. Create a Release PipelineRun using the above info.
+6. Watch Release PipelineRun and update Release `status` with progress and outcome.
+7. If ReleasePlanAdmission specified an `spec.environment`, then do the following:
+   * Copy the Snapshot to the managed workspace
```
[the remark this answers is not in the document]
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md (hunk 18 above)
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md (hunk 19 above)
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md (hunk 7 above)
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md [hunk 21]
```diff
@@ -39,8 +28,95 @@ The diagram below shows the flow of custom resources and the orchestration of pi
 
 ![](../diagrams/release-service/hacbs-release-service-data-flow.jpg)
 
+## Terminology
+
+**Snapshot** -  The custom resource that contains the list of all Components of an Application with their Component Image digests. Once created, the list of Components with their images is immutable. The Integration service updates the status of the resource to reflect the testing outcome.
+**Release Pipeline** -  An instance of a Pipeline that is orchestrated by the Release Service.
+**Bundle** - A tekton bundle (image) that contains Tekton resources
```
[the remark this answers is not in the document]
Comment (scoheb) [reply book/release-service.md]: fixed

File: book/release-service.md [hunk 22]
```diff
@@ -39,8 +28,94 @@ The diagram below shows the flow of custom resources and the orchestration of pi
...
+| SnapshotEnvironmentBinding | When deploying the Snapshot to an Environment            | To update existing SnapshotEnvironmentBinding for a new Snapshot |
+| Release                    | During the lifecycle of an attempt to release a Snapshot | To provide status for the execution of the Release Pipeline      |
+
+### WATCH
+
+| Custom Resource            | When?                                                      | Why?                                                               |
+|----------------------------|------------------------------------------------------------|--------------------------------------------------------------------|
+| Release                    | Always                                                     | To provide an API to trigger a release                             |
+| PipelineRun                | Once the PipelineRun is created                            | To relay the Release PipelineRun status to the Release for viewing |
+| SnapshotEnvironmentBinding | After the SnapshotEnvironmentBinding is created or updated | Ro relay the GitOps deployment status to the Release for viewing   |
```
Comment (johnbieren) [inline book/release-service.md:72]: `s/Ro/To/`

**Review by johnbieren — APPROVED** [review]
lgtm, but I wouldn't merge off of just my approval

**Review by ralphbean — APPROVED** [review]

**Pull request merged**